compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/ClosureDesugar.java

```diff
@@ -212,29 +222,342 @@ public void visit(BLangPackage pkgNode) {
         SymbolEnv pkgEnv = this.symTable.pkgEnvMap.get(pkgNode.symbol);
 
         // Process nodes that are not lambdas
-        pkgNode.topLevelNodes.stream().filter(pkgLevelNode -> !(pkgLevelNode.getKind() == NodeKind.FUNCTION
-                && ((BLangFunction) pkgLevelNode).flagSet.contains(Flag.LAMBDA))).forEach(
-                topLevelNode -> rewrite((BLangNode) topLevelNode, pkgEnv));
+        for (TopLevelNode pkgLevelNode : pkgNode.topLevelNodes) {
+            if (pkgLevelNode.getKind() == NodeKind.FUNCTION
+                    && ((BLangFunction) pkgLevelNode).flagSet.contains(Flag.LAMBDA)) {
+                continue;
+            }
+            if (isObjectConstructorRelated(pkgLevelNode)) {
+                continue;
+            }
+            rewrite((BLangNode) pkgLevelNode, pkgEnv);
+        }
 
         // Reverse the lambdas since in Desugar they are visited from inner to outer lambdas.
         List<BLangLambdaFunction> lambdasCollected = new ArrayList<>(pkgNode.lambdaFunctions);
         Collections.reverse(lambdasCollected);
         pkgNode.lambdaFunctions = new LinkedList<>(lambdasCollected);
-
         while (pkgNode.lambdaFunctions.peek() != null) {
             BLangLambdaFunction lambdaFunction = pkgNode.lambdaFunctions.poll();
             lambdaFunction.function = rewrite(lambdaFunction.function, lambdaFunction.capturedClosureEnv);
         }
 
-        // Update function parameters.
-        pkgNode.functions.forEach(this::updateFunctionParams);
-        pkgNode.typeDefinitions.stream()
-                .filter(typeDef -> typeDef.typeNode.getKind() == NodeKind.RECORD_TYPE)
-                .forEach(this::updateRecordInitFunction);
+        for (BLangClassDefinition classDef : pkgNode.classDefinitions) {
+            if (classDef.isObjectContructorDecl) {
+                updateClassClosureMap(classDef);
+            }
+        }
 
+        // Update function parameters.
+        for (BLangFunction function : pkgNode.functions) {
+            updateFunctionParams(function);
+        }
+        for (BLangTypeDefinition typeDef : pkgNode.typeDefinitions) {
+            if (typeDef.typeNode.getKind() == NodeKind.RECORD_TYPE) {
+                updateRecordInitFunction(typeDef);
+            }
+        }
         result = pkgNode;
     }
 
+    private boolean isObjectConstructorRelated(TopLevelNode pkgLevelNode) {
```

Comment:
Haven't looked into this logic in detail yet, but the closure desugar is basically a desugar phase that does some implementation specific changes to support closures. 

I'm not sure if we should be implementing object constructor specific changes here. We had a similar discussion for the function default value closure change (https://github.com/ballerina-platform/ballerina-lang/pull/31589) and decided  to move that logic out to a separate desugar phase (that runs before the actual desugar to generate the closures and the actual desugar generates calling the closures).

By the time we get here it will be just another closure.

Reply:
Given that this is a blocker for GA and is an internal change, I guess we can go ahead with this for now. But IMO, we need to prioritize moving this out to a separate phase.

Reply:
We also need to make sure we have sufficient tests across modules (via BIR symbol enter), with nested references, etc. We did come across some issues with function default closures.

Reply:
@KRVPerera, do we have an issue to track fixing this?